### ISSUE TITLE
Clear Documentation on Mounting Read Write Folder Between Host and Container

### ISSUE THREAD

**Issue by niccolox** [lxc/lxd#1879]
hi

can we PLEASE get a clear description of a best practice for creating a read write folder between host and container?

Docker has copius documentation and support around Data Volumes and Vagrant has Synced Folders feature which is again a massively important feature

but with LXD I am utterly confused and a little scared by the threads available

https://github.com/lxc/lxd/issues/714
https://github.com/lxc/lxd/issues/592
https://github.com/lxc/lxd/issues/579
https://github.com/lxc/lxd/issues/733
https://github.com/lxc/lxd/issues/850

surely this is a big issue for LXD, having confusing documentation around mounting folders between host and guest could create all sorts of security issues where new LXD users do all sorts of mis-configuration around permissions and groups etc

could we please have some high quality, command line driven documentation?

I'd suggest LXD could do with a Data Volume or Synced Folders type feature with terse and compact and bullet proof command lines

thanks

**Comment by niccolox** [comment]
ok, from docs, relatively easy by making a privileged container

> lxc launch ubuntu:14.04 priv -c security.privileged=true
> lxc config device add priv homedir disk source=/home/$USER path=/home/ubuntu

but files created by container are owned by root, so not sure how useful this really is

would be really useful to get a simple description of command lines for making a safe synced folder setup for an unprivileged container with uid/gid mappings

apache httpd would be a great example

THANKS

**Comment by stgraber** [comment]
There is no simple way of doing it :)

The best way of doing it is by using posix ACLs but that's very far from easy.

There's also nothing LXD can really do about this, it's just how the kernel works.

**Comment by niccolox** [comment]
ok, combined with complexities around LXD networking this is a blocker for me

I am going to have to revert to Trusty and Vagrant and Docker which is hugely disappointing

hopefully over time other users will contribute examples and documentation and perhaps code

I wish I had time to dive into this, I really like the framework, but its too low level for me right now

**Comment by niccolox** [comment]
Implement ACL's with disk devices and add a create_cmd
https://github.com/lxc/lxd/issues/850

**Comment by niccolox** [comment]
having said that... 

something like this below is sufficient, I havent tested this, but this is the kind of thing

> lxc init ....
> sudo btrfs subvolume create /cdata/<hostname>
> sudo setfacl -Rm user:lxd:rwx,default:user:lxd:rwx,user:10000000:rwx,user:10001000:rwx,default:user:10000000:rwx,default:user:10001000:rwx,mask:rwx /cdata/<hostname>
> lxc config device add srv01:<hostname> mntdir disk source=/cdata/<hostname> path=data

**Comment by niccolox** [comment]
is there a LXD plugin or contrib ecosystem? 

could this be done by Ansible?

**Comment by stgraber** [comment]
Ansible could do it indeed. There is no plugin/hook system in LXD itself nor would it be particularly easy in this case as even if we had plugins we wouldn't allow them to extend the standard REST API.

**Comment by stgraber** [comment]
There has been some discussions to add uid/gid shifting as a vfs feature in the kernel which we would certainly then support in LXD. Setting up ACLs automatically or altering the source path permissions in any way while handy may also cause accidental breakage. We prefer to not have LXD do things the user may not expect.

**Comment by niccolox** [comment]
ok, so it comes down to a recipe / howto around LXD for USERS like me

here is another snippet from the Wordpress Roots.io dev community
https://roots.io/linux-containers-lxd-as-an-alternative-to-virtualbox-for-wordpress-development/

> Synced folders
> 
> Vagrant has "synced folders" which shares hosts files to the guest VM via NFS. LXD allows us to mount directories from the host inside our container, which means that we don’t need to use NFS. To make sure that the container can write to the host directory, we need to make the directory world writeable on the host. Issue the following command, substituting the path with the actual path to your Bedrock/Trellis project on the host machine.
> 
> sudo chmod -R go+w /home/user/sites/example.com/site
> I find that this command will not correctly set permissions of the hidden dot files. Make sure .env in particular is world writeable before continuing. Then run the following command to mount the directory within the container, substituting the paths with the correct paths for your project:
> 
> lxc config device add firstcontainer webroot disk path=/srv/www/example.com/current source=/home/user/sites/example.com/site

**Comment by niccolox** [comment]
I add a linkback and comment on the roots.io discourse discussion forum where there seems to be active and ongoing interest

https://discourse.roots.io/t/lxc-lxd-as-an-alternative-to-virtualbox-for-local-development/5149/28

**Comment by Griznah** [comment]
I had quite a few headaches with this myself, so some documentation or semi-official how-to would've been helpful at the time. For now, I'm left with two mountpoints where on the host the uid/gid aren't known, but the shares are rw for containers and hosts at least. For me the security is good enough, but for others it might not.

**Comment by bloodearnest** [comment]
I got this working by creating a custom profile, and adding my user to suduid/gid

https://gist.github.com/bloodearnest/ebf044476e70c4baee59c5000a10f4c8

It basically does the same thing that the old lxc ubuntu template -b option did, except:

1) uid is mapped
2) passwd is not copied, rather, passwordless sudo access is given (since root is not actually uid 0)

I need to check whether cloud init will auto install a users shell if it's not installed, but other than that, it seems to work well.

Be nice if there was an option to set this profile/uidmap up for a user ootb?

**Comment by antitoine** [comment]
I don't know if it's the place to post that, but to work easily I prefer to mount the LXD container in the host machine and mapping the root or other container user with the current host user with the command `bindfs`.
I made a simple script to do it after `lxc start` (and unmount after `lxc stop`) with using custom commands `lxd-start` and `lxd-stop`. After container start, I mount the root of the container in a custom host directory (in my script `/lxd/<container name>/` by default).
The full script is available here : https://github.com/antitoine/lxd-functions

In resume :
```
lxc start mycontainer
sudo bindfs --force-user=1000 --force-group=1000 --create-for-user=165536 --create-for-group=165536 /var/lib/lxd/containers/mycontainer/rootfs /lxd/mycontainer
```
Where `1000` is my host uid and gid and `165536` the container root uid and gid.
In my script I will get automatically the uid and gid of the container (root or a user in the container).

I'm not really happy about this solution, because it's a generic patch that not allow custom mapping by container and `sudo` command is needed to mount the container. It will be great in profile container configuration, or use a native solution that does not use a mounting directory to increase performance.

**Comment by niccolox** [comment]
I think its important there is an official method for this, clearly
described, perhaps embedded as code or plugins or extras or whatever

having people hack around with uid/gid on host and containers is a recipe
for disaster imho

the underlying technology when used properly may be secure, but the more
gray zones, areas left to interpretation the more likely there will be
errors

I read on the roots.io wordpress forum talk of making the mapped folder
world writeable ! what is wrong with this picture?

here is a recent article exploring this
https://www.nccgroup.trust/us/about-us/newsroom-and-events/blog/2016/june/abusing-privileged-and-unprivileged-linux-containers/

On Thu, Jun 2, 2016 at 12:54 AM, Antoine Chabert <notifications@github.com>
wrote:

> I don't know if it's the place to post that, but to work easily I prefer
> to mount the LXD container in the host machine and mapping the root or
> other container user with the current host user with the command bindfs.
> I made a simple script to do it after lxc start (and unmount after lxc
> stop) with using custom commands lxd-start and lxd-stop. After container
> start, I mount the root of the container in a custom host directory (in my
> script /lxd/<container name>/ by default).
> The full script is available here :
> https://github.com/antitoine/lxd-functions
>
> In resume :
>
> lxc start mycontainer
> sudo bindfs --force-user=1000 --force-group=1000 --create-for-user=165536 --create-for-group=165536 /var/lib/lxd/containers/mycontainer/rootfs /lxd/mycontainer
>
> Where 1000 is my host uid and gid and 165536 the container root uid and
> gid.
> In my script I will get automatically the uid and gid of the container
> (root or a user in the container).
>
> I'm not really happy about this solution, because it's a generic patch
> that not allow custom mapping by container and sudo command is needed to
> mount the container. It will be great in profile container configuration,
> or use a native solution that does not use a mounting directory to increase
> performance.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/lxc/lxd/issues/1879#issuecomment-223220954>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AAbspkqfzB8xGe9hrmUV0TJMY7ky4Jk0ks5qHowlgaJpZM4IDv9Z>
> .
>



-- 

-- 
Nicholas Roberts
www.niccolox.org

**Comment by niccolox** [comment]
looking at @bloodearnest  scripts I see lxc.id_map 

```
  # this part maps the special uid/gid in the container to the correct host uid/gid
  raw.lxc: |
    lxc.id_map = u $ID $_UID 1
    lxc.id_map = g $ID $GID 1
```

the official docs do not do justice to the HUGE POTENTIAL USEFULNESS of this

https://linuxcontainers.org/lxc/manpages/man5/lxc.container.conf.5.html

> UID MAPPINGS
> A container can be started in a private user namespace with user and group id mappings. For instance, you can map userid 0 in the container to userid 200000 on the host. The root user in the container will be privileged in the container, but unprivileged on the host. Normally a system container will want a range of ids, so you would map, for instance, user and group ids 0 through 20,000 in the container to the ids 200,000 through 220,000.
> lxc.id_map
> Four values must be provided. First a character, either 'u', or 'g', to specify whether user or group ids are being mapped. Next is the first userid as seen in the user namespace of the container. Next is the userid as seen on the host. Finally, a range indicating the number of consecutive ids to map.
> 

one of the main selling points of Docker is a well documented and well understood Data Volume feature
https://docs.docker.com/engine/userguide/containers/dockervolumes/

same in Vagrant, Synced Folders 
https://www.vagrantup.com/docs/synced-folders/

**Comment by tim-seoss** [comment]
For those watching... the problem is described from a kernel API PoV here:

https://lwn.net/Articles/700422/

TLDR: (Sept 2016) "In conclusion, [...] the problem of accessing container images in unprivileged containers is unsolved at this point, but the broad outlines of potential solutions are taking shape"

"it all needs to be put together in a form that's ready for merging. He doesn't expect that to happen for a few more kernel development cycles"

... the tools which would be necessary for such container image access issues should also solve the upstream kernel API requirements that lxd has when addressing this issue (and hopefully ultimately lead to a reasonable solution being document-able).

In general it's also useful reading to understand how lxd sets things up for unprivileged containers.

**Comment by luislobo** [comment]
@niccolox the problem about making group and other writable, is that any new file does not inherit the security, as far as I have checked. you have to run it again for each new file.

**Comment by bloodearnest** [comment]
Some more up to date methods here: https://insights.ubuntu.com/2016/12/08/mounting-your-home-directory-in-lxd/

I've updated my gist at https://gist.github.com/bloodearnest/ebf044476e70c4baee59c5000a10f4c8

**Comment by devekko** [comment]
big ups to yourself and Canonical for being so responsive 

I haven't test this yet, but this looks exactly the kind of thing

how is performance? I hear from a very experienced devops person that Docker host / guest folder sharing is extremely slow, slower than Virtualbox, which is SLOW

have you done any comparisions on this? 

say I am making a custom theme with node and sass or an elixir app which compiles and a million tiny files get synced between host and guest ? 

how does this compare:

LXD
vs
Docker
vs
NFS
vs
Virtualbox
vs 
Unison

**Comment by bloodearnest** [comment]
On 3 Feb 2017 6:51 pm, "Nicholas Roberts" <notifications@github.com> wrote:

big ups to yourself and Canonical for being so responsive

Well, I work for canonical, but I don't work on lxd particularly, I just
use it a lot, at work and at home

I haven't test this yet, but this looks exactly the kind of thing

how is performance? I hear from a very experienced devops person that
Docker host / guest folder sharing is extremely slow, slower than
Virtualbox, which is SLOW

have you done any comparisions on this?

say I am making a custom theme with node and sass or an elixir app which
compiles and a million tiny files get synced between host and guest ?

There is no filesystem overhead. There may be a slight overhead in user
namespace translation, but only slight. But I would expect near native
performance. But I haven't run any tests. But there is no 'syncing' going
on, save the usual fsync for whatever filesystem the shared mount is on.

how does this compare:

LXD
vs
Docker
vs
NFS
vs
Virtualbox
vs
Unison


Not sure w/o benchmarks

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/lxc/lxd/issues/1879#issuecomment-277329879>, or mute
the thread

**Comment by devekko** [comment]
I should close this, I think this is realistic, and start a new thread at a later stage
